### ISSUE TITLE
Create Logging Implementation That Will Be Used By System Components

### ISSUE THREAD

**Issue by Rogibb111** [Rogibb111/GameBoy-Emulator-Browser#17]
Need to Create a Class that System Components will use to log their data and actions.

The simplest use case for this class would be an instance of it is created in each instance of system component. It would have a key for the class it's being used in, an Id for the instance. The class would then use its log function when it takes any action.

A more sophisticated version would have subject lines when logging for things like data e.g. registers. It would also have logging levels like traditional loggers.

The most sophisticated version would wrap the components instance and inject its own versions of the components data and functions so it could log when things are called and data is changed.

Update: After doing a little research, I have decided to go with the most sophisticated version. The singleton classes should be relatively easy to implement. I will have to come up with a class wrapper for the classes used multiple times. Also, logging the data will be it's own challenge as well (thinking javascript watchers?). 

**Components**
- [ ] MMU 
- [ ] CPU
- [ ] GPU
- [ ] KEY
- [ ] jsGB
- [ ] TIMER

**Models**
- [ ] MBC
- [ ] MemoryBank
- [ ] Address
- [ ] Registers

**Comment by Rogibb111** [comment]
I am creating a logging interface in #56 that will allow me to pick functions and properties to log. Logging functions can be done dynamically based on the test that I have done in this little experiment:

    class TestClass {
        testProp = 'test'
        testFunc = (str: any) => str + "1";
    }
    const testInstance = new TestClass();

    function createLogs(instance: any, functionName: string) {
        const func = instance[functionName];

        instance[functionName] = function logger(...args: any) {
            console.log("Logging before " + functionName);
            const ret = func(args);
            console.log("Logging after " + functionName);
            console.log("Return is: " + ret);
            return ret;
        }
    }

    createLogs(testInstance, 'testFunc');
    const funcRet = testInstance.testFunc('hey');
    console.log("Function Return is " + funcRet);

**Comment by Rogibb111** [comment]
Running into some trickiness with dynamically overriding class definitions. Bottom line is it can't really be done. So my next thought was to extend the multi-instance classes with a base logging class. The problem with this is if I add the array of strings that represent the functions and properties that I would like to log directly to the class, they won't be seen in the logging class's constructor. When typescript compiles classes down to javascript, all property assignments happen in the classes constructor AFTER the base class's super has been called.

**Comment by Rogibb111** [comment]
After doing a bunch more playing in Typescript, I have managed to figure out how I am going to replace properties on the extending class from the base class. Doing a classic assignment using indexing was causing TypeScript errors so I had to get a little creative (typescript hates that the property isn't statically written to the class). Tried using for..in, but that only grabs assigned properties to the class, not functions as well. Luckily I was able leverage some classic JavaScript and made it work via Object.Assign. Example below:

    class LoggingClass {
        loggingFunc(prop: string) {
            Object.assign(this, { [prop]: 2 });
        }
    }


    class TestClass extends LoggingClass{
        testProp: string = "1";

        constructor() {
            super();
            this.loggingFunc('testFunc');
            this.loggingFunc('testProp')
            console.log(this.testFunc)
            console.log(this.testProp);
        }

        testFunc(str: any) { str + "1" }
    }

    new TestClass()

**Comment by Rogibb111** [comment]
So the above solution works great for functions. I can grab all of the functions using Object.assign(), then iterate through the array of functions I need to log, and replace them with proxy logger functions using Object.assign() again. However, this doesn't help me all that much with properties. Primitives can't have logging versions of themselves. After some small research I stumbled onto proxies. These do exactly what I need, but they only operate at an object level, not a property level. Luckily for me, Object.assign() saves my ass again. I can use it to replace the entire this object with a proxy. So my handler function will be some massive thing, but i will be able to intercept and log all gets and sets on properties. Woot example: 

    class LoggingClass {
        loggingFunc(prop: string) {
            Object.assign(this, new Proxy(this, handler2));
        }
    }

    const handler2 = {
        get: function() {
            return "world";
        }
    };  

    class TestClass extends LoggingClass{
        testProp: string = "1";

        constructor() {
            super();
            this.loggingFunc('testProp')
            console.log(this.testProp)
        }
    }

    new TestClass()
// console log reads:
// world

**Comment by Rogibb111** [comment]
I have realized now that their are certain classes (namely Z80) who have sectioned off their properties into separate objects on the class for better organization. The proxy wont be able to handle this without doing some special voodoo magic. After playing around in the TypeScript playground again, I was able to come up with this example that allows me to log sets in nested objects.

`class Test {
    hey = {
        you: 1,
        me: 2
    };
}
const handler = {
    get(obj: any, prop: string): any{
        console.log(prop);
        const propVal = obj[prop];
        console.log(typeof propVal === 'object');
        if (typeof propVal === 'object') {
            return new Proxy(propVal, handler);
        }
        return propVal; 
    },
    set(obj: any, prop: string , val: any) {
        console.log(prop, val);
        obj[prop] = val;
        return true;
    }
}
var prox = new Proxy(new Test(), handler);

console.log(prox.hey.you);
console.log('-----------');
prox.hey.you = 5;
---------------------
[LOG]: hey 
[LOG]: true 
[LOG]: you 
[LOG]: false 
[LOG]: 1 
[LOG]: ----------- 
[LOG]: hey 
[LOG]: true 
[LOG]: you,  5 
`
